files/private-chef-ctl-commands/key-control.rb

```diff
@@ -0,0 +1,255 @@
+      exit 1
+    end
+  end
+
+  def check_client_args(cmd_args, usage_msg)
+    unless cmd_args.length == 3 || \
+      (cmd_args.length == 5 && (cmd_args[3] == "--expiration-date" || cmd_args[3] == "--key-name")) || \
+      (cmd_args.length == 7 && ((cmd_args[3] == "--expiration-date" && cmd_args[5] == "--key-name") || (cmd_args[5] == "--expiration-date" && cmd_args[3] == "--key-name")))
+      STDERR.puts usage_msg
+      STDERR.puts "Usage: Expiration date defaults to infinite. Pass an ISO 8601 fomatted string: YYYY-MM-DDTHH:MM:SS e.g. 2013-12-24T21:00:00"
+      STDERR.puts "Usage: Default name used is the fingerprint of the key passed."
+      exit 1
+    end
+  end
+
+  def add_key(cmd_args, id)
+    # TODO should keyname be optional and default the key fingerprint?
+
+    # this is guaranteed to work because it would have crashed above if not
+    # expiration_date and key_name will be nil if they weren't passed
+    arg_parser = ArgParser.new
+    expiration_date = arg_parser.find_arg_after_identifier(cmd_args, "--expiration-date")
+    key_name        = arg_parser.find_arg_after_identifier(cmd_args, "--key-name")
+    user = cmd_args[0]
+    key = File.read(cmd_args[1])
+
+    unless /-----BEGIN PUBLIC KEY-----/.match(key)
+      STDERR.puts "Error: Invalid public key passed. Key must begin with:"
+      STDERR.puts "Error: -----BEGIN PUBLIC KEY-----"
+      exit 1
+    end
+
+    # if a key_name was not specified, use the fingerprint of the key
+    unless key_name
+      begin
+        openssl_key_object = OpenSSL::PKey::RSA.new(key)
+        data_string = OpenSSL::ASN1::Sequence([
+                                               OpenSSL::ASN1::Integer.new(openssl_key_object.public_key.n),
+                                               OpenSSL::ASN1::Integer.new(openssl_key_object.public_key.e)
+                                              ])
+        key_name = OpenSSL::Digest::SHA1.hexdigest(data_string.to_der).scan(/../).join(':')
+      rescue
+        STDERR.puts "Error: Could not parse fingerprint for public key #{cmd_args[1]}."
+        STDERR.puts "Error: If optional --key-name arg is not specified, then fingerprint of key is used by default as the key_id."
+        exit 1
+      end
+    end
+
+
+    if expiration_date
+      begin
+        Time.parse(expiration_date)
+      rescue
+        STDERR.puts "Error: Invalid date format inserted."
+        STDERR.puts "Error: You inserted #{expiration_date}."
+        STDERR.puts "Error: Please a date in format YYYY-MM-DDTHH:MM:SS e.g. 2013-12-24T21:00:00."
+        exit 1
+      end
+    else
+      expiration_date = "infinity"
+    end
+
+    object = {
+      'id'          => id,
+      'key_id'      => key_name,
+      'public_key'  => key,
+      'key_version' => 0,
+      'created_at'  => Time.now.utc,
+      'expires_at'  => expiration_date
+    }
+
+    begin
+      DB[:keys].insert(object)
```

Comment:
I was giving this some thought today and realized that this will not prevent users /clients from authenticating using their old keys, since we're doing this in an additive way.  Is that the behavior we want?